pkg/controller/pravegacluster/upgrade.go

```diff
@@ -0,0 +1,372 @@
+	}
+
+	// No upgrade in progress
+
+	if p.Spec.Version == p.Status.CurrentVersion {
+		// No intention to upgrade
+		return nil
+	}
+
+	if readyCondition == nil || readyCondition.Status != corev1.ConditionTrue {
+		r.clearUpgradeStatus(p)
+		log.Print("cannot trigger upgrade if there are unready pods")
+		return nil
+	}
+
+	// Need to sync cluster versions
+	log.Printf("syncing cluster version from %s to %s", p.Status.CurrentVersion, p.Spec.Version)
+
+	// Setting target version and condition.
+	// The upgrade process will start on the next reconciliation
+	p.Status.TargetVersion = p.Spec.Version
+	p.Status.SetUpgradingConditionTrue()
+
+	return nil
+}
+
+func (r *ReconcilePravegaCluster) clearUpgradeStatus(p *pravegav1alpha1.PravegaCluster) (err error) {
+	p.Status.SetUpgradingConditionFalse()
+	p.Status.TargetVersion = ""
+	// need to deep copy the status struct, otherwise it will be overridden
+	// when updating the CR below
+	status := p.Status.DeepCopy()
+
+	p.Spec.Version = p.Status.CurrentVersion
+	if err := r.client.Update(context.TODO(), p); err != nil {
+		return err
+	}
+
+	p.Status = *status
+	return nil
+}
+
+func (r *ReconcilePravegaCluster) syncComponentsVersion(p *pravegav1alpha1.PravegaCluster) (err error) {
+	var synced bool
+
+	for _, component := range []componentSyncVersionFun{
+		componentSyncVersionFun{
+			name: "bookkeeper",
+			fun:  r.syncBookkeeperVersion,
+		},
+		componentSyncVersionFun{
+			name: "segmentstore",
+			fun:  r.syncSegmentStoreVersion,
+		},
+		componentSyncVersionFun{
+			name: "controller",
+			fun:  r.syncControllerVersion,
+		},
+	} {
+		synced, err = component.fun(p)
+		if err != nil {
+			return fmt.Errorf("failed to sync %s version. %s", component.name, err)
+		}
+
+		if synced {
+			log.Printf("%s version sync has been completed", component.name)
+		} else {
+			// component version sync is still in progress
+			// Do not continue with the next component until this one is done
+			return nil
+		}
+	}
+
+	// All component versions have been synced
+	p.Status.CurrentVersion = p.Status.TargetVersion
+	return nil
+}
+
+func (r *ReconcilePravegaCluster) syncControllerVersion(p *pravegav1alpha1.PravegaCluster) (synced bool, err error) {
+	deploy := &appsv1.Deployment{}
+	name := util.DeploymentNameForController(p.Name)
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: p.Namespace}, deploy)
+	if err != nil {
+		return false, fmt.Errorf("failed to get deployment (%s): %v", deploy.Name, err)
+	}
+
+	targetImage := fmt.Sprintf("%s:%s", p.Spec.Pravega.ImageRepository, p.Status.TargetVersion)
+
+	if deploy.Spec.Template.Spec.Containers[0].Image != targetImage {
+		// Need to update pod template
+		// This will trigger the rolling upgrade process
+		log.Printf("updating deployment (%s) pod template image to '%s'", deploy.Name, targetImage)
+
+		deploy.Spec.Template = pravega.MakeControllerPodTemplate(p)
+		// deploy.Spec.Template.Spec.Containers[0].Image = targetImage
```

Comment:
Not sure if we need to remove this comment?

Reply:
Yes, it has to be remoed. Good catch!